# Module 6 - Correlation & Linear Regression

In [1]:
import pandas as pd
import numpy as np

## Correlation

In [2]:
Location = "datasets/gradedata.csv"
df = pd.read_csv(Location)

df.head()

fname     lname  gender  age  exercise  hours  grade  \
0   Marcia      Pugh  female   17         3     10   82.4   
1   Kadeem  Morrison    male   18         4      4   78.2   
2     Nash    Powell    male   18         5      9   79.3   
3  Noelani    Wagner  female   14         2      7   83.2   
4  Noelani    Cherry  female   18         4     15   87.4   

                                    address  
0   9253 Richardson Road, Matawan, NJ 07747  
1          33 Spring Dr., Taunton, MA 02780  
2          41 Hill Avenue, Mentor, OH 44060  
3        8839 Marshall St., Miami, FL 33125  
4  8304 Charles Rd., Lewis Center, OH 43035

In [3]:
df.corr()

age  exercise     hours     grade
age       1.000000 -0.003643 -0.017467 -0.007580
exercise -0.003643  1.000000  0.021105  0.161286
hours    -0.017467  0.021105  1.000000  0.801955
grade    -0.007580  0.161286  0.801955  1.000000

## Linear Regression

In [4]:
import statsmodels.formula.api as smf

In [5]:
result = smf.ols('grade ~ age + exercise + hours', data=df).fit()

In [6]:
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  grade   R-squared:                       0.664
Model:                            OLS   Adj. R-squared:                  0.664
Method:                 Least Squares   F-statistic:                     1315.
Date:                Wed, 16 Jan 2019   Prob (F-statistic):               0.00
Time:                        15:18:23   Log-Likelihood:                -6300.7
No. Observations:                2000   AIC:                         1.261e+04
Df Residuals:                    1996   BIC:                         1.263e+04
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     57.8704      1.321     43.804      0.000      55.279      60.461
age            0.0397      0.075      0.532      0.595      -0.107       0.186
exercise       0.9893      0.089     11.131      0.000       0.815       1.164
hours          1.9165      0.031     61.564      0.000       1.855       1.978
==============================================================================
Omnibus:                      321.187   Durbin-Watson:                   2.047
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             2196.187
Skew:                          -0.567   Prob(JB):                         0.00
Kurtosis:                       8.007   Cond. No.                         213.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

With age, exercise, and hours being 0, your starting grade is likely to be around 57.87

In [7]:
#remove age from regression, not very correlated
result = smf.ols('grade ~ exercise + hours', data=df).fit()
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  grade   R-squared:                       0.664
Model:                            OLS   Adj. R-squared:                  0.664
Method:                 Least Squares   F-statistic:                     1973.
Date:                Wed, 16 Jan 2019   Prob (F-statistic):               0.00
Time:                        15:19:05   Log-Likelihood:                -6300.8
No. Observations:                2000   AIC:                         1.261e+04
Df Residuals:                    1997   BIC:                         1.262e+04
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     58.5316      0.447    130.828      0.000      57.654      59.409
exercise       0.9892      0.089     11.131      0.000       0.815       1.163
hours          1.9162      0.031     61.575      0.000       1.855       1.977
==============================================================================
Omnibus:                      318.721   Durbin-Watson:                   2.048
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             2158.000
Skew:                          -0.564   Prob(JB):                         0.00
Kurtosis:                       7.962   Cond. No.                         43.2
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [8]:
#set coefficient to 0
result = smf.ols(formula='grade ~ age + exercise + hours - 1', data=df).fit()
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  grade   R-squared:                       0.991
Model:                            OLS   Adj. R-squared:                  0.991
Method:                 Least Squares   F-statistic:                 7.284e+04
Date:                Wed, 16 Jan 2019   Prob (F-statistic):               0.00
Time:                        15:19:41   Log-Likelihood:                -6974.3
No. Observations:                2000   AIC:                         1.395e+04
Df Residuals:                    1997   BIC:                         1.397e+04
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
age            3.1129      0.035     88.030      0.000       3.044       3.182
exercise       1.7659      0.122     14.482      0.000       1.527       2.005
hours          2.2860      0.042     54.486      0.000       2.204       2.368
==============================================================================
Omnibus:                      131.221   Durbin-Watson:                   2.006
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              403.367
Skew:                          -0.301   Prob(JB):                     2.57e-88
Kurtosis:                       5.116   Cond. No.                         14.2
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### Your Turn

Run a correlation and regresssion on the dataset below. What can you tell from the data?

In [9]:
Location = "datasets/datasets/tamiami.csv"

df = pd.read_csv(Location)
df.head()

FileNotFoundError: File b'datasets/datasets/tamiami.csv' does not exist

In [ ]:
columns = ['location', 'sales', 'employees', 'restaurants', 'foodcarts', 'price']

#change column names for readability
df.columns = columns
df.head()